### PR TITLE
Adding a basic test for lossless.

### PR DESCRIPTION
**Pull request by vrabaud**
This will be used when YCgCo-R is added next.

### CODE REVIEW

**Comment by joedrago** [comment]
Please run clang-format on all files in your commit. Files outside of the gtest dir do not use Google's format style.

File: apps/shared/avifutil.c
Comment (wantehchang) [inline apps/shared/avifutil.c:227]: Is it possible to modify the `avifInputReadImage` function in avifenc.c to call this new function?

File: apps/shared/avifutil.c
Comment (wantehchang) [inline apps/shared/avifutil.c:229]: Please run clang-format. Use the clang-format command line in https://github.com/AOMediaCodec/libavif/blob/main/README.md.

File: tests/are_images_equal.cc
Comment (wantehchang) [inline tests/are_images_equal.cc:23]: Check the return value of `avifReadImage` (two occurrences)

File: tests/gtest/aviftest_helpers.h
Comment (wantehchang) [inline tests/gtest/aviftest_helpers.h:37]: Document the `ignore_alpha` parameter.

File: tests/are_images_equal.cc
Comment (wantehchang) [inline tests/are_images_equal.cc:1]: Our current policy allows C++ code only in the tests/gtest directory. Since this file doesn't use GoogleTest, it may seem inappropriate to put this file in tests/gtest.

One solution is to move the `libavif::testutil::AreImagesEqual()` function to apps/shared/avifutil.h. We will need to convert it to a C function, and therefore won't be able to specify a default argument for the `ignore_alpha` parameter.

Another solution is to move this file to tests/gtest.

Note: We can also write the lossless test using GoogleTest. Then we don't need to add this file.

File: tests/are_images_equal.cc
Comment (wantehchang) [inline tests/are_images_equal.cc:19]: Nit: Check if `decoded0` and `decoded1` are null pointers.

**Review by wantehchang — CHANGES_REQUESTED** [review]

File: apps/shared/avifutil.c
Comment (y-guyon) [reply apps/shared/avifutil.c:229]: [The clang-format-check GitHub action](https://github.com/vrabaud/libavif/runs/7217915080?check_suite_focus=true#step:4:10) did not catch the issue. There might be some configuration error, maybe because of the `**` patterns?

`/entrypoint.sh: line 44: local: 'apps/shared/iccjpeg.h': invalid variable name for name reference` also puzzles me.

File: tests/test_cmd.sh
Comment (wantehchang) [inline tests/test_cmd.sh:56]: IMPORTANT: The `are_images_equal` program compares the two images after converting them to YUV. For lossless compression, it seems that we should compare the images in RGB form.

File: apps/shared/avifutil.c
Comment (vrabaud) [reply apps/shared/avifutil.c:227]: Done

File: apps/shared/avifutil.c
Comment (vrabaud) [reply apps/shared/avifutil.c:229]: dir/**.h does not take subdirectories.

Comment (vrabaud) [reply apps/shared/avifutil.c:229]: Done

File: apps/shared/avifutil.h
Comment (vrabaud) [reply apps/shared/avifutil.h:49]: Done

File: tests/are_images_equal.cc
Comment (vrabaud) [reply tests/are_images_equal.cc:1]: The test is about testing avifenc/avifdec, not the algorithms. Moving it to the gtest folder.

File: tests/are_images_equal.cc
Comment (vrabaud) [reply tests/are_images_equal.cc:17]: Done

File: tests/are_images_equal.cc
Comment (vrabaud) [reply tests/are_images_equal.cc:19]: Done

File: tests/are_images_equal.cc
Comment (vrabaud) [reply tests/are_images_equal.cc:23]: Done

File: tests/are_images_equal.cc
Comment (vrabaud) [reply tests/are_images_equal.cc:19]: Indeed: I set the MC to identity so that it stays in RGB for PNG or YUV for JPEG.

File: tests/test_cmd.sh
Comment (vrabaud) [reply tests/test_cmd.sh:56]: Indeed, done.

File: tests/gtest/are_images_equal.cc
Comment (wantehchang) [inline tests/gtest/are_images_equal.cc:43]: It would be better to use the exit status 1 for "images are not equal", so that this exit status is independent of the number of errors we want to differentiate by exit status.

I think it is fine to use a single exit status for all errors. Your call.

File: tests/test_cmd.sh
Comment (wantehchang) [inline tests/test_cmd.sh:56]: It would be good to explain why alpha is ignored in this test.

File: tests/gtest/are_images_equal.cc
Comment (wantehchang) [inline tests/gtest/are_images_equal.cc:15]: Nit: Change `./are_image_equal` to `argv[0]`:

```
        << "Wrong argument: " << argv[0] << " file1 file2 ignore_alpha_flag"
```

File: tests/gtest/are_images_equal.cc
Comment (wantehchang) [inline tests/gtest/are_images_equal.cc:27]: It seems that we should set kRequestedDepth to 0. 0 means "request the bit depth closest to the bit depth of the input file".

The `avifImage` cannot use the exact bit depth of the input fie when the input file is 16-bit PNG; we are forced to use 12-bit AVIF in that case.

For lossless test, if the bit depth of the `avifImage` is not equal to the bit depth of the input file, there is potentially loss from the `avifReadImage()` call.

File: tests/CMakeLists.txt
Comment (wantehchang) [inline tests/CMakeLists.txt:105]: `AVIF_BUILD_APPS` means "build the executables in the apps/ directory".

Note: We can move are_images_equal.cc to the apps/ directory, but we will need to convert it to a C file, and we will also need to move `libavif::testutil::AreImagesEqual` to apps/shared/avifutil.{h,c} and convert it to a C function.

**Review by wantehchang — CHANGES_REQUESTED** [review]

File: apps/avifenc.c
Comment (vrabaud) [reply apps/avifenc.c:288]: I'd rather keep the current format which is what is done above too: exit as soon as you can.

File: tests/CMakeLists.txt
Comment (vrabaud) [reply tests/CMakeLists.txt:45]: `AVIF_BUILD_TESTS` should not be used in this file as `AVIF_BUILD_TESTS` already controls whether we `add_subdirectory` this folder.
`are_images_equal` is, for now, only used to test avifenc/avifdec. Therefore, it is only used if the apps are built, and therefore is `AVIF_BUILD_APPS` is set.

If you want, we can decide that `AVIF_BUILD_TESTS` sets `AVIF_BUILD_APPS` to ON.

File: tests/CMakeLists.txt
Comment (vrabaud) [reply tests/CMakeLists.txt:105]: I am adding a test to test the app so I indeed need to check for `AVIF_BUILD_APPS`

File: tests/gtest/are_images_equal.cc
Comment (vrabaud) [reply tests/gtest/are_images_equal.cc:29]: I wrote it this way because it is easier to understand to me, and shorter. With one glance, I know that it tests for 0 and 1, I do not have to think about what is the last value.

File: tests/test_cmd.sh
Comment (vrabaud) [reply tests/test_cmd.sh:56]: I am not ignoring it anymore. It will be ignored for a future test comparing with a JPEG: JPEG doe snot have alpha loaded, while PNG has it loaded (with full opacity though).

File: tests/gtest/are_images_equal.cc
Comment (vrabaud) [reply tests/gtest/are_images_equal.cc:32]: Done

File: tests/test_cmd.sh
Comment (wantehchang) [reply tests/test_cmd.sh]: I think I understand the issue now. We should enhance `libavif::testutil::AreImagesEqual` to treat an absent alpha plane and a fully-opaque alpha plane as equal. We can do that in a follow-up CL. The `avifImageIsOpaque()` function in libavif/src/write.c could be useful.

Also, generating a PNG with a fully-opaque alpha plane when the image has no alpha is a know issue: https://github.com/AOMediaCodec/libavif/issues/787. That issue should be straightforward to fix.

File: tests/gtest/are_images_equal.cc
Comment (wantehchang) [inline tests/gtest/are_images_equal.cc:23]: We should return an exit status >= 2 here because this is an error. (1 now means "images are not equal".)

File: tests/gtest/are_images_equal.cc
Comment (wantehchang) [inline tests/gtest/are_images_equal.cc:36]: We should return an exit status >= 2 here because this is an error. (1 now means "images are not equal".)

File: tests/test_cmd.sh
Comment (wantehchang) [inline tests/test_cmd.sh:58]: It would be better to check if the exit status is equal to 1, right? We can check the `$?` variable.

**Review by wantehchang — CHANGES_REQUESTED** [review]
Just some simple changes, otherwise this looks good.

File: tests/test_cmd.sh
Comment (vrabaud) [reply tests/test_cmd.sh]: Indeed, the flag should probably be something like: ignore_alpha_if_opaque

File: tests/gtest/are_images_equal.cc
Comment (jzern) [inline tests/gtest/are_images_equal.cc:34]: NULL could be nullptr, though I don't know how consistent the c++ code is at the moment

File: tests/test_cmd.sh
Comment (jzern) [inline tests/test_cmd.sh:60]: join this with the line above and prefer [[ ]] there

File: tests/test_cmd.sh
Comment (jzern) [inline tests/test_cmd.sh:62]: indent can be 2 to match the rest of the file

File: tests/test_cmd.sh
Comment (jzern) [inline tests/test_cmd.sh:67]: This isn't strictly necessary given the echo should succeed, but fine either way.

File: tests/gtest/are_images_equal.cc
Comment (wantehchang) [reply tests/gtest/are_images_equal.cc:34]: Since this is a new file, let's use `nullptr` instead of `NULL`.

**Review by wantehchang — COMMENTED** [review]
Vincent: LGTM. I will approve this pull request after you have addressed James's comments.

File: tests/test_cmd.sh
Comment (wantehchang) [inline tests/test_cmd.sh:57]: James, Vincent: Is it possible to avoid temporarily setting `+e` for this test?

Comment (jzern) [reply tests/test_cmd.sh:57]: if "${ARE_IMAGES_EQUAL}" "${TESTDATA_DIR}"/kodim23_yuv420_8bpc.y4m "${DECODED_FILE}" 0; then
...
fi

Comment (wantehchang) [reply tests/test_cmd.sh:57]: How do we check if the exit status of that `${ARE_IMAGES_EQUAL}` command is equal to 1?

Exit status 1 means "images are not equal". Exit status 2 means "the command failed with an error". So this test expects the exit status to be equal to 1, rather than any nonzero value.

Comment (vrabaud) [reply tests/test_cmd.sh:57]: I set it as final: failing tests will be executed last.

**Review by wantehchang — APPROVED** [review]
LGTM.

**Pull request merged**
